### PR TITLE
Fixes, asyncCall() consistency and execution tests

### PR DESCRIPTION
**Pull request by camilbancioiu**
This PR adds the following:

* Changed `ExecuteOnSameContext()` and `ExecuteOnDestContext()` to support recursive calls.
* Added max instance limit (currently non-configurable).
* Added tests for recursive calls of `ExecuteOnSameContext()` and `ExecuteOnDestContext()`.
* Improved stack-related behavior of the contexts, to support bifurcations (no longer than 1 state), required for recursive calls. This also allows:
  * storage access from calls higher up on the stack
  * reverting the value transfer of a failed `ExecuteOn(Same|Dest)Context()` call
* Changed how the main methods in `execution.go` return a `vmOutput` in case of error, with proper `ReturnCode` resolution; the `outputContext` now handles this everywhere. Also reorganized errors defined by Arwen.
* Changed `asyncCall()` to return a `vmOutput` that is always usable, containing changes performed by the parent, child and `callBack()`, regardless of which one fails (except the first step of the parent, which cannot be recovered). 
* Added tests for `asyncCall()` when either the child or the `callBack()` exit with failure.
* Added extra tests in general.

TODO:
* configurable max instance limit
* blockchain hooks for "contract in same shard?" and "address is smartcontract?", to be used in `asyncCall()`
* other TODOs and fixes
* SC API tests in `elrondei.go`
* refactor tests in `execution_test.go`, to remove duplication

### CODE REVIEW

File: arwen/host/execution.go [hunk 1]
```diff
@@ -189,33 +177,108 @@ func (host *vmHost) doRunSmartContractCall(input *vmcommon.ContractCallInput) (v
...
+		runtime.PopSetActiveState()
+
+		return vmOutput, err
+	}
+
+	err = host.execute(input)
+	if err != nil {
+		// Execution failed: restore contexts as if the execution didn't happen,
+		// but first create a vmOutput to capture the error.
+		vmOutput := output.CreateVMOutputInCaseOfError(err)
```
Comment (sasurobert) [inline arwen/host/execution.go:216]: make a function out of these 4 lines as they are the same as for transfer error.

Comment (sasurobert) [reply arwen/host/execution.go:216]: you could put to defer if err -> code

File: arwen/host/execution.go [hunk 2]
```diff
@@ -189,33 +177,108 @@ func (host *vmHost) doRunSmartContractCall(input *vmcommon.ContractCallInput) (v
...
 	runtime.InitStateFromContractCallInput(input)
-	err := host.execute(input)
 
-	return err
+	// Perform a value transfer to the called SC. If the execution fails, this
+	// transfer will not persist.
+	err := output.Transfer(input.RecipientAddr, input.CallerAddr, 0, input.CallValue, nil)
+	if err != nil {
+		// Execution failed: restore contexts as if the execution didn't happen.
+		bigInt.PopSetActiveState()
```
Comment (sasurobert) [inline arwen/host/execution.go:258]: same as above. duplicated code plus use defer if err ->

File: test/contracts/exec-dest-ctx-recursive-parent/exec-dest-ctx-recursive-parent.arwen [hunk 3]
```diff
@@ -0,0 +1 @@
...1002903f08880800037031002400240200242ff0183500d0041002005417f6a3a00b88a808000024041002903908980800041a08980800041f089808000200041106a4110410141b48a80800041b88a808000108980808000220841014b0d000240024020080e020001000b200041046a41002d00ba888080003a0000200041002800b688808000360200200041041081808080000c030b200041046a41002d00bf888080003a0000200041002800bb88808000360200200041041081808080000c020b200041046a41002d00c4888080003a0000200041002800c088808000360200200041041081808080000c010b41c08980800041002802e489808000108a808080001a0b200041c0006a2480808080000b0bbf0203004180080b8101586b65794e4e4e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e005876616c75654e4e4e005866696e6973684e4e4e0073756363006661696c00756e6b6e00000000000000000000000077726f6e67206e756d626572206f6620617267756d656e7473000000000000006368696c6443616c6c73506172656e7400004190090ba801a08601000000000000000000000000006368696c6453432e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e726563757273697665497465726174696f6e426967436f756e7465722e2e2e2e000000005800000000000000000000000000000000000000000000000000000000000000000000000000000000000005726563757273697665497465726174696f6e436f756e7465722e2e2e2e2e2e2e00000000010000000041b80a0b0100@0500
```
Comment (andreibancioiu) [inline test/contracts/exec-dest-ctx-recursive-parent/exec-dest-ctx-recursive-parent.arwen:1]: Just a note, the new version of erdpy won't generate `*.hex.arwen` files anymore.

File: arwen/contexts/bigint.go [hunk 4]
```diff
@@ -4,65 +4,77 @@ import (
 	"math/big"
 )
 
+type bigIntMap map[int32]*big.Int
+
 type bigIntContext struct {
-	mappedValues map[int32]*big.Int
-	stateStack   []*bigIntContext
+	values     map[int32]*big.Int
```
[the remark this answers is not in the document]
Comment (camilbancioiu) [reply arwen/contexts/bigint.go:10]: Indeed, changed.

File: arwen/contexts/bigint.go [hunk 5]
```diff
@@ -4,65 +4,77 @@ import (
...
-	context.mappedValues = make(map[int32]*big.Int)
+	context.values = make(map[int32]*big.Int)
 }
 
 func (context *bigIntContext) PushState() {
-	newState := &bigIntContext{
-		mappedValues: context.mappedValues,
-	}
-
+	newState := context.duplicateState()
```
[the remark this answers is not in the document]
Comment (camilbancioiu) [reply arwen/contexts/bigint.go:29]: Renamed.

File: arwen/contexts/metering.go [hunk 6]
```diff
@@ -108,6 +108,10 @@ func (context *meteringContext) UnlockGasIfAsyncStep() {
 	context.gasLockedForAsyncStep = 0
 }
 
+func (context *meteringContext) GetGasLockedForAsyncStep() uint64 {
```
[the remark this answers is not in the document]
Comment (camilbancioiu) [reply arwen/contexts/metering.go:111]: I thought about it as well. The caller must decide how much gas is needed for the `callBack()`. There will be a new PR for this (will likely include changes in `elrond-go` and probably `elrond-vm-common` as well, to store the user-specified gas lock).

https://elrondnetwork.atlassian.net/browse/EN-6071

File: arwen/contexts/output.go [hunk 7]
```diff
@@ -190,15 +214,62 @@ func (context *outputContext) DeployCode(input arwen.CodeDeployInput) {
...
 		ReturnMessage: message,
 	}
 }
 
+func (context *outputContext) resolveReturnCodeFromError(err error) vmcommon.ReturnCode {
+	if err != nil {
+		if errors.Is(err, arwen.ErrSignalError) {
+			return vmcommon.UserError
+		}
+		if err == arwen.ErrFuncNotFound {
```
[the remark this answers is not in the document]
Comment (camilbancioiu) [reply arwen/contexts/output.go:242]: Changed to use `errors.Is` for all cases, and added blank lines between `if` conditions for ordered error chains, for clarity.

I wouldn't know how to use a `switch` with `errors.Is`, but will change if actually possible.

File: arwen/contexts/output.go [hunk 8]
```diff
@@ -237,7 +308,7 @@ func mergeOutputAccounts(
 		leftAccount.BalanceDelta = big.NewInt(0)
 	}
 	if rightAccount.BalanceDelta != nil {
-		leftAccount.BalanceDelta = big.NewInt(0).Add(leftAccount.BalanceDelta, rightAccount.BalanceDelta)
+		leftAccount.BalanceDelta = rightAccount.BalanceDelta
```
[the remark this answers is not in the document]
Comment (camilbancioiu) [reply arwen/contexts/output.go:311]: Previously, `Add` was required indeed. This PR changes the meaning of the state stacks, from "partial states that can be merged" to actual "full checkpoints", with limited branching and revert (with a minor exception - storage updates, which still support merging). 

More info on how state stacking has changed: a good example is `executeOnSameContext()`, which must either finish successfuly, or all its changes reverted in case of failure, including any value transfers. This was solved by changing the contexts to hold an "active state" and a "state stack". Pushing on the state stack means copying the active state to the stack, while continuing to work with the active state. 

When `executeOnSameContext()` runs, it will modify the "active state" (with storage updates, `BalanceDelta` etc.). If it returns with success, the active state is considered "good", and the previous copy (at the top of the stack) is deleted with `PopDiscard()`. If it returns with failure, the active state is considered "wrong" and it is replaced with the copy from the top of the stack, with `PopSetActiveState()`.

The source line highlighted here in `output.go` is only reached by `executeOnDestContext()` on a successful return, when it must return to the previous SC on the stack. This requires a merge between the last known state of the previous SC (from the stack) with the _new_ active state produced by `executeOnDestContext()`. A simple replacement with `PopSetActiveState()` would not have been sufficient, like for `executeOnSameContext()`, because `executeOnDestContext()` copies it for the new SC (so that balances and nonces are visible to the new SC), but then censors it, so that the new SC appears to be running in isolation (almost everything is erased, except storage updates, which aren't accessible anyway). 

In conclusion, assignment is required here now, because `leftAccount.BalanceDelta` and `rightAccount.BalanceDelta` were initially identical, but `rightAccount.BalanceDelta` is the new value after `executeOnDestContext()`. 

This is tested by `TestExecution_ExecuteOnDestContext_Recursive_Mutual_SCs()`, with two SCs calling each other recursively using `executeOnDestContext()`.

While the current stacking concept makes sense and appears to work, it is not easy enough to understand from the code directly, and could be improved. I'm open to any idea.

File: arwen/contexts/storage.go [hunk 9]
```diff
@@ -112,7 +119,7 @@ func (context *storageContext) SetStorage(key []byte, value []byte) int32 {
 		return int32(arwen.StorageAdded)
 	}
 	if bytes.Equal(value, zero) {
-		freeGas := metering.GasSchedule().BaseOperationCost.StorePerByte * uint64(len(oldValue))
+		freeGas := metering.GasSchedule().BaseOperationCost.ReleasePerByte * uint64(len(oldValue))
 		metering.FreeGas(freeGas)
 		return int32(arwen.StorageDeleted)
 	}
```
[the remark this answers is not in the document]
Comment (camilbancioiu) [reply arwen/contexts/storage.go:125]: Right, this needs to be handled. But should we allow a "free write to storage" if the new value is shorter than the old value? 

We definitely must release the gas for the difference, but we could still consume gas for persisting the shorter new value, because it will be written to storage. But overall, doing this would result in a lower total cost for `storageStore()` (i.e. `gasForWritingNewValue` - `gasReleasedByDifference`).

This would cover the case where a SC persisting 1000 values of storage at full cost, each 100 bytes in size, but later just overwriting them with a single different byte. Because the new and old values are the same length, this can be abused with very little gas cost, but higher resource cost for the node.

File: arwen/host/asyncCall.go [hunk 10]
```diff
@@ -26,16 +25,14 @@ func (host *vmHost) handleAsyncCallBreakpoint() error {
 	}
 
 	destinationVMOutput, err := host.ExecuteOnDestContext(destinationCallInput)
```
[the remark this answers is not in the document]
Comment (camilbancioiu) [reply arwen/host/asyncCall.go:27]: Agreed, changed.

File: arwen/host/asyncCall.go [hunk 11]
```diff
@@ -26,16 +25,14 @@ func (host *vmHost) handleAsyncCallBreakpoint() error {
...
-	}
 
-	callbackCallInput, err := host.createCallbackContractCallInput(destinationVMOutput)
+	callbackCallInput, err := host.createCallbackContractCallInput(destinationVMOutput, err)
 	if err != nil {
 		return err
 	}
 
-	_, err = host.ExecuteOnDestContext(callbackCallInput)
+	callbackVMOutput, err := host.ExecuteOnDestContext(callbackCallInput)
```
[the remark this answers is not in the document]
Comment (camilbancioiu) [reply arwen/host/asyncCall.go:34]: Changed.

File: arwen/host/asyncCall.go [hunk 12]
```diff
@@ -93,12 +108,13 @@ func (host *vmHost) createDestinationContractCallInput() (*vmcommon.ContractCall
...
-			CallType:    vmcommon.AsynchronousCall,
-			GasPrice:    runtime.GetVMInput().GasPrice,
-			GasProvided: gasLimit,
+			CallerAddr:    sender,
+			Arguments:     arguments,
+			CallValue:     big.NewInt(0).SetBytes(asyncCallInfo.ValueBytes),
+			CallType:      vmcommon.AsynchronousCall,
+			GasPrice:      runtime.GetVMInput().GasPrice,
+			GasProvided:   gasLimit,
+			CurrentTxHash: runtime.GetTxHash(),
```
[the remark this answers is not in the document]
Comment (camilbancioiu) [reply arwen/host/asyncCall.go:117]: Added.

File: arwen/host/asyncCall.go [hunk 13]
```diff
@@ -130,12 +153,13 @@ func (host *vmHost) createCallbackContractCallInput(destinationVMOutput *vmcommo
...
-			CallType:    vmcommon.AsynchronousCallBack,
-			GasPrice:    runtime.GetVMInput().GasPrice,
-			GasProvided: gasLimit,
+			CallerAddr:    sender,
+			Arguments:     arguments,
+			CallValue:     big.NewInt(0),
+			CallType:      vmcommon.AsynchronousCallBack,
+			GasPrice:      runtime.GetVMInput().GasPrice,
+			GasProvided:   gasLimit,
+			CurrentTxHash: runtime.GetTxHash(),
```
[the remark this answers is not in the document]
Comment (camilbancioiu) [reply arwen/host/asyncCall.go:162]: Added.

File: arwen/host/execution.go [hunk 14]
```diff
@@ -37,14 +36,14 @@ func (host *vmHost) doRunSmartContractCreate(input *vmcommon.ContractCreateInput
...
-	vmOutput, err = host.performCodeDeploy(codeDeployInput)
-	return
+	vmOutput, err := host.performCodeDeploy(codeDeployInput)
+	if err != nil {
+		return output.CreateVMOutputInCaseOfError(err)
+	}
+	return vmOutput
 }
 
 func (host *vmHost) onExitDirectCreateOrCall(err error, vmOutput *vmcommon.VMOutput) *vmcommon.VMOutput {
```
[the remark this answers is not in the document]
Comment (camilbancioiu) [reply arwen/host/execution.go:46]: Forgot to delete this function. Clean-up is still needed, and this function was replaced with `defer` statements in `doRunSmartContract***()`.

But it's a better idea to have a function that cleans this up. Added `host.Clean()` and called it with `defer` where required.

File: arwen/host/execution.go (hunk 1 above)
Comment (camilbancioiu) [reply arwen/host/execution.go:216]: Rewrote this using `defer` and handled in a separate function, depending on `err`.

File: arwen/host/execution.go (hunk 2 above)
Comment (camilbancioiu) [reply arwen/host/execution.go:258]: Rewrote this using `defer` and handled in a separate function, depending on `err`.

File: arwen/contexts/bigint.go [hunk 15]
```diff
@@ -4,65 +4,77 @@ import (
...
 	context.stateStack = append(context.stateStack, newState)
 }
 
-func (context *bigIntContext) PopState() {
+func (context *bigIntContext) PopSetActiveState() {
```
[the remark this answers is not in the document]
Comment (camilbancioiu) [reply arwen/contexts/bigint.go:33]: :(

File: arwen/contexts/bigint.go [hunk 16]
```diff
@@ -4,65 +4,77 @@ import (
...
+	newState := make(bigIntMap, len(context.values))
+	for id, bigInt := range context.values {
+		newState[id] = big.NewInt(0).Set(bigInt)
+	}
+	return newState
 }
 
 func (context *bigIntContext) Put(value int64) int32 {
-	newIndex := int32(len(context.mappedValues))
+	newIndex := int32(len(context.values))
```
[the remark this answers is not in the document]
Comment (camilbancioiu) [reply arwen/contexts/bigint.go:59]: Right, renamed to `handle` and `newHandle`.

File: arwen/contexts/output.go [hunk 17]
```diff
@@ -71,15 +80,30 @@ func (context *outputContext) PopState() {
...
+func (context *outputContext) PopDiscard() {
+	stateStackLen := len(context.stateStack)
+	context.stateStack = context.stateStack[:stateStackLen-1]
+}
+
 func (context *outputContext) ClearStateStack() {
 	context.stateStack = make([]*vmcommon.VMOutput, 0)
 }
 
+func (context *outputContext) CensorVMOutput() {
```
[the remark this answers is not in the document]
Comment (camilbancioiu) [reply arwen/contexts/output.go:98]: Added comment.

File: arwen/contexts/output.go [hunk 18]
```diff
@@ -190,15 +214,62 @@ func (context *outputContext) DeployCode(input arwen.CodeDeployInput) {
...
+		GasRemaining:  metering.GetGasLockedForAsyncStep(),
 		GasRefund:     big.NewInt(0),
-		ReturnCode:    errCode,
+		ReturnCode:    returnCode,
 		ReturnMessage: message,
 	}
 }
 
+func (context *outputContext) resolveReturnCodeFromError(err error) vmcommon.ReturnCode {
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (camilbancioiu) [reply arwen/contexts/output.go:238]: Fixed. I keep making this mistake.

File: arwen/contexts/runtime.go [hunk 19]
```diff
@@ -64,15 +66,25 @@ func (context *runtimeContext) InitState() {
 }
 
 func (context *runtimeContext) CreateWasmerInstance(contract []byte, gasLimit uint64) error {
-	var err error
-	context.instance, err = wasmer.NewMeteredInstance(contract, gasLimit)
+	if context.RunningInstancesCount() >= context.maxWasmerInstances {
```
[the remark this answers is not in the document]
Comment (camilbancioiu) [reply arwen/contexts/runtime.go:69]: Renamed to `StartWasmerInstance()`, to also suggest that now it "runs" in the background.

File: arwen/contexts/storage.go [hunk 20]
```diff
@@ -63,9 +68,11 @@ func (context *storageContext) GetStorage(key []byte) []byte {
...
 
 	value, _ := context.blockChainHook.GetStorageData(context.address, key)
-	storageUpdates[string(key)] = &vmcommon.StorageUpdate{
-		Offset: key,
-		Data:   value,
+	if value != nil {
```
[the remark this answers is not in the document]
Comment (camilbancioiu) [reply arwen/contexts/storage.go:71]: No, this simply brings a value from the Trie into the `StorageUpdates`, which acts as a RW cache. If a value is brought from the `BlockchainHook` (Trie) just for reading, it will stay unmodified in `StorageUpdates`. 

If an entry is deleted, it is added in `StorageUpdates` with `[]byte{}`. See `storageContext.SetStorage()`

File: arwen/host/execution.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (camilbancioiu) [reply arwen/host/execution.go:46]: Removed, it was forgotten.

File: arwen/host/execution.go [hunk 21]
```diff
@@ -55,14 +54,7 @@ func (host *vmHost) overrideVMOutputIfError(err error, vmOutput *vmcommon.VMOutp
...
 
-	var message string
-	if err == arwen.ErrSignalError {
-		message = output.ReturnMessage()
-	} else {
-		message = err.Error()
-	}
-
-	return output.CreateVMOutputInCaseOfError(output.ReturnCode(), message)
+	return output.CreateVMOutputInCaseOfError(err)
```
[the remark this answers is not in the document]
Comment (camilbancioiu) [reply arwen/host/execution.go:57]: Removed.

File: arwen/contexts/runtime_test.go [hunk 22]
```diff
@@ -212,7 +212,11 @@ func TestRuntimeContext_PushPopState(t *testing.T) {
 	runtimeContext.SetVMInput(nil)
 	runtimeContext.SetReadOnly(true)
 
-	runtimeContext.PopState()
+	require.Equal(t, []byte("dummy"), runtimeContext.GetSCAddress())
```
[the remark this answers is not in the document]
Comment (camilbancioiu) [reply arwen/contexts/runtime_test.go:215]: Fixed.

File: arwen/host/execution.go [hunk 23]
```diff
@@ -394,28 +458,34 @@ func (host *vmHost) EthereumCallData() []byte {
...
-			return err
+	runtime := host.Runtime()
+	init := runtime.GetInitFunction()
+	if init == nil {
+		return nil
+	}
+
+	_, err := init()
+	if err != nil {
+		breakpointValue := runtime.GetRuntimeBreakpointValue()
```
[the remark this answers is not in the document]
Comment (camilbancioiu) [reply arwen/host/execution.go]: Done.

File: arwen/contexts/storage.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (sasurobert) [reply arwen/contexts/storage.go:125]: you already took the cost of calling the save Method :)

**Comment by camilbancioiu** [comment]
> `TestExecution_ManyDeployments` seems to be slower now (compared to master branch).

I expanded a test SC and forgot it was used by `TestExecution_ManyDeployments()`. Fixed now, it's even 3x faster than `master`.

File: arwen/contexts/storage.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (camilbancioiu) [reply arwen/contexts/storage.go:125]: Indeed. Changed to only consume the extra difference, or to release gas if the new value is shorter.
